dev-portal/src/services/api-catalog.js

```diff
@@ -1,144 +1,207 @@
-import { getApiGatewayClient } from './api'
-export let subscriptions
-let catalog
+import { apiGatewayClient } from './api'
+import { store } from './state'
+
+/* Catalog and API Utils */
+
+// wrap the catalogPromiseCache
+export const updateCatalogAndApisList = ((() => {
+  let catalogPromiseCache
+
+  return function(bustCache = false) {
+    let catalogOrPromise = store.catalog.length ? store.catalog : catalogPromiseCache
+    if (!bustCache && catalogOrPromise) return Promise.resolve(catalogOrPromise)
+
+    return catalogPromiseCache = apiGatewayClient()
+      .then(apiGatewayClient => apiGatewayClient.get('/catalog', {}, {}, {}))
+      .then(({ data = [] }) => {
+        // attach usage plan data to each api
+        store.catalog = data.map(usagePlan => {
+          usagePlan.apis = usagePlan.apis.map(api => {
+            // remove the apis from the cloned usagePlan so we don't go circular
+            api.usagePlan = cloneJson(usagePlan)
+            delete api.usagePlan.apis
+            return api
+          })
+
+          return usagePlan
+        })
 
-export function getCatalog() {
-  if (catalog) return Promise.resolve(catalog)
+        updateSubscriptionStatus()
 
-  return fetchCatalog()
-  .then(({data}) => {
-    catalog = data
-    return data
-  })
-}
+        store.apiList = store.catalog.reduce((acc, usagePlan) => acc.concat(usagePlan.apis), [])
 
-export function getApi(apiId) {
-  return getCatalog()
+        return store.catalog
+      })
+  }
+})())
+
+// will return the first api if no id provided
+export function getApi(apiId, useFirst) {
+  return updateCatalogAndApisList() // we're only waiting so that we can make sure this has been set
   .then(() => {
-    let _api
 
-    catalog.forEach(c => {
-      if (_api) return
-      _api = c.apis.find(a => a.id === apiId)
-    })
+    console.log(store.apiList)
 
-    return _api
+    let catalog = store.catalog
+    if (catalog && catalog.length) {
+      let api
+      
+      catalog.find(c => (api = c.apis.find(a => a.id === apiId)))
+      
+      return api
+    }
   })
 }
 
-export function fetchCatalog() {
-  return getApiGatewayClient().then(apiGatewayClient => {
-    return apiGatewayClient.get('/catalog', {}, {}, {})
-  })
+/**
+ * Updates the viewed api to the api id provided. If that api ID
+ */
+export function updateApi(apiId) {
+  return getApi(apiId)
+    .then(api => store.api = api)
```

Comment:
Is `store.api` the... currently selected api? Unclear...

Reply:
Yes it is. What would make it clearer?